### PR TITLE
Feature: Implement client message cache expiration and memory limitation

### PR DESCRIPTION
**Pull request by omani**
### Proposed changes in this pull request
Proposal for #827.

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [x] Read the [CONTRIBUTION guidelines](https://github.com/nknorg/nkn#contributing).
- [x] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [x] Added proper documentation where ever applicable (in code and README.md).
- [x] Code has been written according to [NKN-Golang-Style-Guide](https://github.com/nknorg/nkn/wiki/NKN-Golang-Style-Guide)

### Extra information
- tested against race conditions
- test cases follow after general approval/review.

### CODE REVIEW

File: api/websocket/messagebuffer/messagebuffer.go [hunk 1]
```diff
@@ -1,39 +1,218 @@
...
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return true
+	}
+
+	return true
+}
+
+func (m *Buffer) Purge() {
```
Comment (yilunzhang) [inline api/websocket/messagebuffer/messagebuffer.go:77]: Purge has O(N) time complexity, which is too high as a regular cleaning job

Comment (omani) [reply api/websocket/messagebuffer/messagebuffer.go:77]: correct. I have my trouble with purging this way. I am thinking about a better way to solve this. I left `Purge()` in there to demonstrate/test that an iteration over the whole map is needed to keep a clean cache state.

another way would be an irregular O(n) by doing the cleanup periodically and sacrificing accuracy.
yet another way would be to purge only when cache is full.

any ideas are welcome.

File: api/websocket/messagebuffer/messagebuffer.go [hunk 2]
```diff
@@ -1,39 +1,218 @@
...
-	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messageBuffer.buffer[clientIDStr] = append(messageBuffer.buffer[clientIDStr], msg)
+func (m *Buffer) AddMessage(clientID string, relay *pb.Relay) error {
+	if relay.MaxHoldingSeconds <= 0 {
+		return nil
+	}
+
+	m.Purge()
```
Comment (yilunzhang) [inline api/websocket/messagebuffer/messagebuffer.go:100]: Calling Purge in AddMessage have two unwanted results:
1. AddMessage has O(N) time complexity
2. If no new messages arrive, old messages will never be removed

File: api/websocket/messagebuffer/messagebuffer.go [hunk 3]
```diff
@@ -1,39 +1,218 @@
...
 }
 
-// PopMessages reads and clears all messages of a client
-func (messageBuffer *MessageBuffer) PopMessages(clientID []byte) []*pb.Relay {
-	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messages := messageBuffer.buffer[clientIDStr]
-	messageBuffer.buffer[clientIDStr] = nil
+func (m *Buffer) PopMessages(clientID string) []*pb.Relay {
```
Comment (yilunzhang) [inline api/websocket/messagebuffer/messagebuffer.go:109]: You are trying to implement a heap right? Why not using the built in heap? https://pkg.go.dev/container/heap

File: api/websocket/messagebuffer/messagebuffer.go [hunk 4]
```diff
@@ -1,39 +1,218 @@
...
+		},
+	}
+
+	m.queue[clientID] = append(m.queue[clientID], node)
+
+	par := getParent(next)
+	cur := next
+
+	for (par >= 0) && (q[par].isAfter(node)) {
+		m.queue[clientID][par], m.queue[clientID][cur] = q[cur], q[par]
```
Comment (yilunzhang) [inline api/websocket/messagebuffer/messagebuffer.go:156]: This seems not right. len(q) is not increased after append (but len(m.queue[clientID]) is), so you are accessing q[len(q)]

File: api/websocket/messagebuffer/messagebuffer.go (hunk 3 above)
Comment (omani) [reply api/websocket/messagebuffer/messagebuffer.go:109]: yes. this is a heap implementation. I didnt think of container/heap.

File: api/websocket/messagebuffer/messagebuffer.go (hunk 1 above)
Comment (yilunzhang) [reply api/websocket/messagebuffer/messagebuffer.go:77]: You can use another heap (sorted by expiration time) for fast (O(log N)) purge. The caveat is that when removing an element from one heap, you need to remove it from the other one as well.

**Comment by omani** [comment]
switched to container/heap. commit not squashed to reflect PR discussion.

**Comment by omani** [comment]
b.size <= 1000 is for testing.

**Comment by omani** [comment]
we need to discuss the purging.

**Comment by yilunzhang** [comment]
One way to purge efficiently is using another heap sorted by expiration time. This way each purge call has O(log N) time complexity per element being removed. This heap can be shared by all message recipients. The caveat is that when removing an element from one heap, you need to remove it from the other one as well.

**Comment by omani** [comment]
> One way to purge efficiently is using another heap sorted by expiration time. This way each purge call has O(log N) time complexity per element being removed. This heap can be shared by all message recipients. The caveat is that when removing an element from one heap, you need to remove it from the other one as well.

as you can see I have a second heap. but it is just a minheap of clients. when I maxheap for expirationtime, dont we remove clients that a newer but have longer maxholdingseconds? instead, this second heap is just removing the oldest client.

**Comment by yilunzhang** [comment]
Yeah as you mentioned, the second heap isn't really helping us achieve fast purge

**Comment by yilunzhang** [comment]
Closing this PR as it's implemented in #876

**Pull request closed without merging**